[Substrate.io Migration] Tutorials -> Build a Parachain -> Transfer Assets with XCM (#30)
polkadot-developers/polkadot-docs: develop/parachain-devs/interoperability/transfer-assets-with-xcm.md

To teleport assets between the two chains, follow these steps:

1. Navigate to the **Extrinsics** tab of the relay chain in the Polkadot.js Apps interface
    1. Select the **Developer** tab from the sidebar
    2. Choose the **Extrinsics** option

        ![](/images/develop/parachain-devs/interoperability/transfer-assets-with-xcm/transfer-assets-with-xcm-1.webp)

2. Select the **xcmPallet** and the **limitedTeleportAssets** extrinsic
    1. Choose the **xcmPallet** from the list of pallets
    2. Select the **limitedTeleportAssets** extrinsic from the extrinsics available

        ![](/images/develop/parachain-devs/interoperability/transfer-assets-with-xcm/transfer-assets-with-xcm-2.webp)

3. Fill in the required parameters for the extrinsic

>     1. Enter the **dest** multilocation of the destination chain. In this case, since the destination chain is Asset Hub (ID `1000`), its Multilocation from the Relay Chain perspective is:

Reviewer: Yes, it should be "relay chain"

    6. Click on **Submit Transaction** to execute the extrinsic

        ![](/images/develop/parachain-devs/interoperability/transfer-assets-with-xcm/transfer-assets-with-xcm-3.webp)

4. After submitting the transaction, you can check the status of the teleportation by navigating to the **Events** tab on the **Explorer** page for the relay chain in the Polkadot.js Apps interface. If the XCM message is successfully sent, you will see the event `xcmPallet.Sent` in the list of events

    ![](/images/develop/parachain-devs/interoperability/transfer-assets-with-xcm/transfer-assets-with-xcm-4.webp)

5. To verify the teleportation on the Asset Hub parachain, navigate to the **Events** section of the **Explorer** page. You should see two events, **`messageQueu.Processed`** and **`balances.Transfer`**

    ![](/images/develop/parachain-devs/interoperability/transfer-assets-with-xcm/transfer-assets-with-xcm-5.webp)

These events indicate that the teleportation of assets was successful. You can now verify the balance of the beneficiary account on the Asset Hub parachain to confirm the transfer.

## Reserve Asset Transfer

> When consensus systems lack sufficient trust for direct asset teleportation, they can leverage a trusted third-party reserve (such as Asset Hub) to hold the actual assets. To track ownership of these reserve-held assets, both the source and destination chains typically create derivative tokens. Each chain maintains a dedicated account, known as a sovereign account, on the reserve to manage its asset holdings.

Reviewer 2: Elaborated a bit more

```suggestion
It is a way of transferring assets between two parachains (or consensus systems) that don't trust each other by using a third system (like the Asset Hub system chain) they both trust, called the reserve. To track ownership of the reserve-held assets, both the source and destination chains typically create derivative tokens. Each chain maintains a dedicated account, known as a sovereign account, on the reserve to manage its asset holdings. 

Typically, the sender chain burns a certain amount of derivative tokens, signals the reserve to move the real assets from its sovereign account to the destination chain's sovereign account, and then signals the recipient to mint the right amount of derivatives.
```

Reviewer: ```suggestion
A reserve asset transfer transfers assets between two parachains (or consensus systems) that don't trust each other by using a third system (like the Asset Hub system chain) they both trust, called the reserve. Both the source and destination chains typically create derivative tokens to track ownership of the reserve-held assets. Each chain maintains a dedicated account, known as a sovereign account, on the reserve to manage its asset holdings. 
	
Typically, the sender chain burns a certain amount of derivative tokens, signals the reserve to move the real assets from its sovereign account to the destination chain's sovereign account, and then signals the recipient to mint the right amount of derivatives.
```

Grammarly improvements for the above suggestion

    6. Click on **Submit Transaction** to execute the extrinsic

        ![](/images/develop/parachain-devs/interoperability/transfer-assets-with-xcm/transfer-assets-with-xcm-3.webp)

4. After submitting the transaction, you can check the status of the teleportation by navigating to the **Events** tab on the **Explorer** page for the relay chain in the Polkadot.js Apps interface. If the XCM message is successfully sent, you will see the event `xcmPallet.Sent` in the list of events

    ![](/images/develop/parachain-devs/interoperability/transfer-assets-with-xcm/transfer-assets-with-xcm-4.webp)

5. To verify the teleportation on the Asset Hub system chain, navigate to the **Events** section of the **Explorer** page. You should see two events, **`messageQueu.Processed`** and **`balances.Transfer`**

    ![](/images/develop/parachain-devs/interoperability/transfer-assets-with-xcm/transfer-assets-with-xcm-5.webp)

These events indicate that the teleportation of assets was successful. You can now verify the balance of the beneficiary account on the Asset Hub parachain to confirm the transfer.

## Reserve Asset Transfer

> It is a way of transferring assets between two parachains (or consensus systems) that don't trust each other by using a third system (like the Asset Hub system chain) they both trust, called the reserve. To track ownership of the reserve-held assets, both the source and destination chains typically create derivative tokens. Each chain maintains a dedicated account, known as a sovereign account, on the reserve to manage its asset holdings. 

Reviewer: ```suggestion
A reserve asset transfer transfers assets between two parachains (or consensus systems) that don't trust each other by using a third system (like the Asset Hub system chain) they both trust, called the reserve. Both the source and destination chains typically create derivative tokens to track ownership of the reserve-held assets. Each chain maintains a dedicated account, known as a sovereign account, on the reserve to manage its asset holdings. 
```

Grammarly edits

---
title: Transfer Assets with XCM
description: This tutorial will guide you through the process of transferring assets between parachains using Cross-Chain Message (XCM) instructions.
---

# Transfer Assets with XCM

## Introduction

> In the tutorial on [Open Message Passing Channels](TODO:update-path){target=\_blank} you learned how to establish a two-way communication channel between chains by sending messages to the relay chain. You can apply a similar approach to send messages that enable a local chain to manage an account on a remote chain. This tutorial will teach you how to transfer assets between chains using Cross-Chain Message (XCM) instructions.

Reviewer 3: You can add the link to `Open Message Passing Channels` because its already on master

To transfer assets between two chains, you have the options of [Teleporting Assets](https://wiki.polkadot.network/docs/learn-teleport){target=\_blank} or performing a [Reserve Asset Transfer](https://wiki.polkadot.network/docs/learn/xcm/journey/transfers-reserve){target=\_blank}. This tutorial will cover both methods.

## Prerequisites

Before you begin, verify the following:

- You have a relay chain and a parachain running to test the XCM instructions
- You have opened the message passing channel to allow communication between the two chains

## Teleporting Assets

Asset teleportation enables the transfer of digital assets, including fungible and non-fungible tokens, across different parachains. This process allows the transferred assets to function as if they were native to the receiving chain. To simulate the teleportation of assets, this tutorial showcases the transfer of fungible tokens between the Polkadot relay chain and the Asset Hub system parachain.

> To replicate the scenario, this guide uses [Chopsticks](https://github.com/AcalaNetwork/chopsticks){target=\_blank} to interact with the relay chain and the Asset Hub parachain. To setup the same environment, run the following command:

Reviewer 3: I would just mention that this example teleports funds from Polkadot to Asset Hub. And mention that they can replicate it locally using Chopsticks by following the example in https://github.com/polkadot-developers/polkadot-docs/blob/master/develop/application-devs/tooling/chopsticks/overview.md#xcm-testing

5. To verify the teleportation on the Asset Hub parachain, navigate to the **Events** section. You should see two events, **`messageQueu.Processed`** and **`balances.Transfer`**

    ![](/images/develop/parachain-devs/interoperability/transfer-assets-with-xcm/transfer-assets-with-xcm-5.webp)

These events indicate that the teleportation of assets was successful. You can now verify the balance of the beneficiary account on the Asset Hub parachain to confirm the transfer.

## Reserve Asset Transfer

When consensus systems lack sufficient trust for direct asset teleportation, they can leverage a trusted third-party reserve (such as Asset Hub) to hold the actual assets. To track ownership of these reserve-held assets, both the source and destination chains typically create derivative tokens. Each chain maintains a dedicated account, known as a sovereign account, on the reserve to manage its asset holdings.

To simulate a reserve asset transfer, this tutorial demonstrates that funcionality between the Polkadot relay chain and the Astar parachain. This tutorial uses Chopstick to simulate the interaction between the two chains. To set up the environment, run the following command:

> ```bash

Reviewer 3: Same here, just mention which chains participate in the transfer and link to chopsticks or whatever

chopsticks xcm \
--r polkadot \

> --p polkadot-asset-hub \

Reviewer 3: I think asset hub is not needed

---
title: Transfer Assets with XCM
description: This tutorial will guide you through the process of transferring assets between parachains using Cross-Chain Message (XCM) in a Polkadot parachain.
---

# Transfer Assets with XCM

## Introduction

In the [Open Message Passing Channels](TODO:update-path){target=\_blank} tutorial, you saw how to open a two-way communication channel between chains by sending messages to the relay chain. You can use a similar strategy to send messages that allow a local chain to manage an account on a remote chain. In this tutorial, parachain B transfers assets into the sovereign account on the relay chain for parachain A.

The outcome for this tutorial is similar to using the `transfer` function from the balances pallet, except in this case the transfer is initiated by a parachain and demonstrates how the holding register is used when executing the `WithdrawAsset` and `DepositAssetXCM` instructions.

## Prerequisites

Before you begin, verify the following:

- You have the Sudo pallet available for both local parachains to use
- You have opened the message passing channel to allow communication between parachain B and parachain A

> - You have set up a parachain test network using Zombienet or a local relay chain using the `rococo-local` chain specification. This tutorial uses the following Zombienet configuration file:

Reviewer 3: The prerequisite should be something like you have a network with a relay chain and at least two parachains. I would not include the zombienet config used to get this working

    ```toml
    [relaychain]
    default_command = "polkadot"
    chain = "rococo-local"

    [[relaychain.nodes]]
    name = "alice"
    validator = true
    ws_port = 9944

    [[relaychain.nodes]]
    name = "bob"
    validator = true
    ws_port = 9955

    [[parachains]]
    id = 1000

    [parachains.collator]
    name = "parachainA"
    ws_port = 9988
    command = "parachain-template-node"
    args = [ "-l=xcm=trace" ]

    [[parachains]]
    id = 1001

    [parachains.collator]
    name = "parachainB"
    ws_port = 9989
    command = "parachain-template-node"
    args = [ "-l=xcm=trace" ]

    [settings]
    timeout = 1000
    ```

    !!! note

>         Due to a well-known issue with Zombienet and HRMP channels, the opening of the HRMP channels between parachains has been forced by the relay chain through the `sudo` pallet for this tutorial. For further information about this workaround, please refer to [https://github.com/paritytech/polkadot-sdk/pull/1616#issuecomment-1727194584](https://github.com/paritytech/polkadot-sdk/pull/1616#issuecomment-1727194584){target=\_blank}

Reviewer 3: You can use the name of the issue or something similar instead of the full URL

>         Due to a well-known issue with Zombienet and HRMP channels, the opening of the HRMP channels between parachains has been forced by the relay chain through the `sudo` pallet for this tutorial. For further information about this workaround, please refer to [https://github.com/paritytech/polkadot-sdk/pull/1616#issuecomment-1727194584](https://github.com/paritytech/polkadot-sdk/pull/1616#issuecomment-1727194584){target=\_blank}

Reviewer 3: You can use the name of the issue or something similar instead of the full URL

> ## Configure XCM Instructions

Reviewer 3: This title is slightly confusing

> To illustrate the interaction between the two chains, in the following example, parachain B sends XCM instructions to deposit assets into an account on parachain A.

Reviewer 3: In this section, you don't mention anything related to deposit assets. You are just configuring the `dest` parameter. The following sections explain how to add instructions to the `message` field, but this is not explained

1. Navigate to the **Extrinsics** tab in the Polkadot.Js Apps interface
      1. Select the **Developer** tab
      2. Choose the **Extrinsics** option from the dropdown menu

        ![](/images/tutorials/polkadot-sdk/build-a-parachain/transfer-assets-with-xcm/transfer-assets-with-xcm-1.webp)

2. Use the **sudo** pallet to send the XCM instructions
      1. Select the **sudo** pallet from the dropdown menu

>       2. Choose the **sudo(call)** extrinsic to use the Sudo pallet to execute privileged transactions

Reviewer 3: The name of the extrinsic is just `sudo`. I would not include `call` because, in other examples, we don't include the parameters
```suggestion
      2. Choose the **sudo** extrinsic to use the Sudo pallet to execute privileged transactions
```

        ![](/images/tutorials/polkadot-sdk/build-a-parachain/transfer-assets-with-xcm/transfer-assets-with-xcm-2.webp)

3. Select the **polkadotXcm** pallet to send the XCM instructions
      1. Select **polkadotXcm** pallet from the dropdown menu
      2. Select **send(dest, message)** extrinsic to send a Cross-Chain Message (XCM) to the destination chain
      3. Specify the **dest** parameter to indicate the destination chain for the message. The destination chain is the relay chain, so the value destination looks like this:
            ```javascript
            { "parents": 1, "interior": "Here" }
            ```
      4. Specify the **message** parameter to indicate the XCM version for the message. In this case, the value is `V2`

>       5. Click on **+ Add item** to construct the message to be executed

Reviewer 3: The step `e` in the image is over **Submit Transaction** and not over **Add Item**

        ![](/images/tutorials/polkadot-sdk/build-a-parachain/transfer-assets-with-xcm/transfer-assets-with-xcm-3.webp)

> Now that you have configured the XCM instructions, you can proceed to execute the message between parachains.

Reviewer 3: You cannot execute the message yet because you still need to build it.

      1. Click on **+ Add item** to add the instruction to the message
      2. Select **BuyExecution** from the dropdown menu
      3. Set the **id** parameter to the location of the asset to be withdrawn. In this case, the asset is located in parachain B, so the value looks like this:

          ```json
          { 
            "parents": 0, 
            "interior": "Here"
          }
          ```

      4. Set the **fun** parameter to `true` to identify the asset as a fungible asset
      5. Specify the **fungible** parameter to indicate the total fungible assets to withdraw. For example, this tutorial uses `12000000000000`
      6. Select `Unlimited` to skip setting a weight limit for this instruction on the **weightLimit** parameter

        ![](/images/tutorials/polkadot-sdk/build-a-parachain/transfer-assets-with-xcm/transfer-assets-with-xcm-6.webp)

## DepositAsset Instruction

To deposit assets after fees from the holding register into a specific account:

1. Prepare the [DepositAsset](https://github.com/polkadot-fellows/xcm-format#depositasset){target=\_blank} as the third instruction for this message

>       1. Click on **+ Add item** to add the instruction to the message

Reviewer 3: ```suggestion
      1. Click on **+ Add item** to add a new instruction to the message
```

      2. Select **DepositAsset** from the dropdown menu
      3. For the **asset** parameter, set `Wild` to allow an unspecified number of assets to be deposited. Then, set `All` for the *wild* parameter to allow all of the remaining assets after fees are paid to be deposited
      4. For the **maxAssets** parameter, set `1` as the maximum number of unique assets to remove from the holding register for the deposit. In this tutorial, there's only one asset instance available to be removed
      5. For the **beneficiary** parameter, set the destination account for the deposit. In this case, the account is located in parachain A, so the value looks like this:

         ```json
         {
           "parents": 0,
           "interior": {
             "X1": {
               "junction": {
                 "AccountId32": {
                   "network": "Any",
                   "id": "5DEy8ViuBxh8z8hxeNyBhj4y5Mz2sNybv2aPqDHBh7EsyRAz"
               }
             }
           }
         }
         ```

        ![](/images/tutorials/polkadot-sdk/build-a-parachain/transfer-assets-with-xcm/transfer-assets-with-xcm-7.webp)

>         !!! note

Reviewer 3: I don't think this is relevant for the example

            If you want to deposit the assets into the sovereign account for parachain A, you could specify the beneficiary using the following settings:
                  ```json
                  {
                    "parents": 0,
                    "interior": {
                      "X1": {
                        "junction": "Parachain",
                        "parachainIndex": 1000
                      }
                    }
                  }
                  ```

> After you configure all of the XCM instructions to be executed, you're ready to submit the transaction.

Reviewer 3: I would like a final screenshot with an arrow pointing to the **Submit Transaction** button

## Submit and Check the Transaction

> After submitting the transaction, you can check the status of the transaction in the **Explorer** tab on the parachain B interface. If the transaction is successful, you should see the `polkadotXcm.Sent` event in the **Events** tab.

Reviewer 3: The `sent` extrinsic was successfully executed, but I think the actual transfer could have failed in the remote chain
